examples/lookup/lookup.cpp

```diff
@@ -162,44 +267,136 @@ int main(int argc, char ** argv){
         llama_batch_clear(batch_tgt);
         llama_batch_add(batch_tgt, draft[0], n_past, { 0 }, true);
 
-        // generate n_pred tokens through prompt lookup
+        auto get_token = [](const std::vector<llama_token> inp, const std::vector<llama_token> draft, const size_t i) -> llama_token {
+            // Helper function to get a token from the combined, speculative sequence of inp and draft.
+            return i < inp.size() ? inp[i] : draft[1 + i - inp.size()];
+        };
+
         auto prompt_lookup = [&]() -> void {
+            // Generate up to n_draft additional tokens through prompt lookup.
+            // The draft is aborted early if there is no suitable token candidate to continue the draft.
+            // At the beginning of this function the draft already contains a single token sampled from the model.
             const int inp_size = inp.size();
-            for (int ngram_size = ngram_max ; ngram_size > ngram_min; --ngram_size){
-                const llama_token * ngram = &inp[inp_size - ngram_size];
-
-                for (int i = 0; i <= (int) inp_size - (ngram_size * 2); ++i) {
-                    bool match = true;
-                    for (int j = 0; j < ngram_size; ++j) {
-                        if (inp[i + j] != ngram[j]) {
-                            match = false;
-                            break;
+
+            while ((int) draft.size()-1 < n_draft) {
+                bool draft_success = false;
+
+                const int static_ngram_start = inp_size-2 + draft.size()-1;
+                uint64_t static_ngram = get_token(inp, draft, static_ngram_start);
+                for (int j = static_ngram_start; j < static_ngram_start + 2; ++j) {
+                    const uint64_t ngram_part = get_token(inp, draft, j);
+                    static_ngram <<= 16;
```

Comment:
This seems incorrect - it's calling `get_token(inp, draft, static_ngram_start);` twice: one before the loop on line 285 and one in the first loop iter on line 287

Also, printing the resulting `static_ngram`, the major 16-bits are always 0

Reply:
Ok I see

>  The impact of the bug should only affect n-grams of size 4 but those are not very useful anyways because 3 already overfits.

My intuitive understanding is that the larger the n-gram that we can match in the corpus - the better. For example, for code completion use cases this is most certainly true due to the repetitive nature of code

Reply:
When it comes to generating code or highly structured data (e.g. JSON) or summarizing/extracting information, large n-gram matches are much more frequent compared to free-form text completion. My main desire here is that the potential implementation for n-gram caches that we will implement in `llama` will be general enough and not have limitations on the n-gram size

Reply:
I mean, the main reason I'm encoding the tokens as 64 bit integers is that `uint64_t` is a hashable type. If we were to write a custom hashing function we could go up to whatever length we want.